pkg/kubelet/kubelet_node_status.go

```diff
@@ -429,6 +429,85 @@ func (kl *Kubelet) initialNode(ctx context.Context) (*v1.Node, error) {
 	return node, nil
 }
 
+// fastNodeStatusUpdate is a "lightweight" version of syncNodeStatus which doesn't hit the
+// apiserver except for the final run, to be called by fastStatusUpdateOnce in each loop.
+// It holds the same lock as syncNodeStatus and is thread-safe when called concurrently with
+// syncNodeStatus. Its return value indicates whether the loop running it should exit
+// (final run), and it also sets kl.containerRuntimeReadyExpected.
+func (kl *Kubelet) fastNodeStatusUpdate(ctx context.Context, timeout bool) (completed bool) {
```

Comment:
I feel like this code is unnecessarily complicated.

> // fastNodeStatusUpdate is a "lightweight" version of syncNodeStatus which doesn't hit the
> // apiserver except for the final run

If we want to achieve this, can we just do:
1) More frequent readiness check in an inner loop;
2) Less frequent node status patch in an outer loop?

I guess we just need to separate the node readiness population and the node status update, so that we can populate the readiness and check internally, but update apiserver less frequently?

Or in the worst case, a true of false to `syncNodeStatus` to specify whether it is a "dry-run" sync or not, so that we get the expected node readiness, but doesn't necessarily need to hit the apiserver until ready?

Reply:
> Or in the worst case, a true of false to `syncNodeStatus` to specify whether it is a "dry-run" sync or not, so that we get the new node readiness, but doesn't necessarily need to hit the apiserver until ready?

A call to `syncNodeStatus` always *reads* from the apiserver even if we don't patch it. I tried to change it in a way that it reads from lister instead, but this results in dirtier code https://github.com/kubernetes/kubernetes/commit/203989aca82096d5734c627d0fdb3669bc07a587, so I finally went on the current approach.

Reply:
Thanks for the explanation!

The motivation makes sense to me, but I'm not sure whether the code resulted is cleaner than https://github.com/kubernetes/kubernetes/pull/113188. :P It is mainly a matter of those locks, which are quite error prone.

Please make sure to find a way to clean them up, if we decide to do that in a follow up PR.

Reply:
yeah in this PR I'm more concerned about making things safe. Some locks are unnecessary and some others are shared that might cause confusion. I've tried to clean them up, but decided to go with a follow-up PR to make review easier.

Reply:
With #113188 we got to change all syncNodeStatus callers including the tests. It's more difficult to prove the correctness of the change unlike #113466 where all existing tests stay intact and are still passing.

Reply:
@Random-Liu unhold if this resolved your concerns?